performance/metrics.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package performance
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	metricsNamespace     = "loadtest"
+	metricsSubSystemHTTP = "http"
+	metricsSubSystemWS   = "websocket"
+)
+
+type UserEntityMetrics struct {
+	HTTPRequestTimes     prometheus.Histogram
+	HTTPErrors           *prometheus.CounterVec
+	HTTPTimeouts         prometheus.Counter
+	WebSocketConnections prometheus.Gauge
+}
+
+type Metrics struct {
+	registry  *prometheus.Registry
+	ueMetrics UserEntityMetrics
+}
+
+func NewMetrics() *Metrics {
+	var m Metrics
+	m.registry = prometheus.NewRegistry()
+
+	m.ueMetrics.HTTPRequestTimes = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: metricsNamespace,
+		Subsystem: metricsSubSystemHTTP,
+		Name:      "request_time",
+		Help:      "The time taken to execute client requests.",
+	})
+	m.registry.MustRegister(m.ueMetrics.HTTPRequestTimes)
+
+	m.ueMetrics.HTTPErrors = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: metricsNamespace,
+		Subsystem: metricsSubSystemHTTP,
+		Name:      "errors_total",
+		Help:      "The total number of HTTP client errors.",
+	},
+		[]string{"path", "method", "status_code"},
+	)
+	m.registry.MustRegister(m.ueMetrics.HTTPErrors)
+
+	m.ueMetrics.HTTPTimeouts = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: metricsNamespace,
+		Subsystem: metricsSubSystemHTTP,
+		Name:      "timeouts_total",
+		Help:      "The total number of HTTP client timeouts.",
+	})
+	m.registry.MustRegister(m.ueMetrics.HTTPTimeouts)
+
+	m.ueMetrics.WebSocketConnections = prometheus.NewGauge(prometheus.GaugeOpts{
```

Comment:
Are we looking for a difference between the server side tracking of websocket connections vs client side ? Shouldn't both be the same ?

Reply:
They should. This is the way to prove it. 
Mainly I thought that when it comes to potential connectivity problems I'd rather rely on client side metrics so I added it to actually see how both metrics behave in case of issues.